File: Data Structures & Algorithms/count-connected-components/submission-7.cpp
class Solution {
public:
    int countComponents(int n, vector<vector<int>>& edges) {
        unordered_map<int, vector<int>> adj_list(n);
        for (const vector<int>& edge : edges) {
            int n1 = edge[0], n2 = edge[1];
            adj_list[n1].push_back(n2);
            adj_list[n2].push_back(n1);
        }

        unordered_set<int> visited;
        function<void(int vert)> dfs = [&](int vert) -> void {
            if (visited.contains(vert)) return;

            visited.insert(vert);
            for (int connected_node : adj_list[vert]) dfs(connected_node);
        };

        int out = 0;
        for (int vert = 0; vert < n; vert++) {
            if (!visited.contains(vert)) {
                out++;
                dfs(vert);
            }
        }
        return out;
    }
};

File: Data Structures & Algorithms/valid-tree/submission-7.cs
public class Solution {
    public bool ValidTree(int n, int[][] edges) {
        if (edges.Length != n - 1) return false;

        List<int>[] adjList = new List<int>[n];
        for (int i = 0; i < n; i++) adjList[i] = new List<int>();

        foreach (int[] edge in edges) {
            adjList[edge[0]].Add(edge[1]);
            adjList[edge[1]].Add(edge[0]);
        }

        HashSet<int> visited = new HashSet<int>();
        bool Dfs(int node, int parent) {
            if (visited.Contains(node)) return false;

            visited.Add(node);
            foreach (int neighbor in adjList[node]) {
                if (neighbor == parent) continue;
                if (!Dfs(neighbor, node)) return false;
            }

            return true;
        }

        return Dfs(0, -1) && visited.Count == n;
    }
}
